### ISSUE TITLE
getting "Can't verify CSRF token authenticity" even if csrf toke in present 

### ISSUE THREAD

**Issue by rtcoms** [plataformatec/devise#2734]
I'm doing simple authentication (without ajax or api) and getting error for csrf authenticity token. I'm using latest version of rails and devise

check the post request below

```
 Parameters: {"utf8"=>"✓",    
 "authenticity_token"=>"SJnGhXXUXjncnPhCdg3muV2GYCA8CX2LVFV78pqddD4=", "user"=> 
{"email"=>"a@a.com", "password"=>"[FILTERED]", "remember_me"=>"0"},
"commit"=>"Sign in"}

Can't verify CSRF token authenticity
 User Load (0.4ms)  SELECT "users".* FROM "users" WHERE "users"."email" =  
 'a@a.com' LIMIT 1
(0.1ms)  begin transaction
SQL (0.4ms)  UPDATE "users" SET "last_sign_in_at" = ?, "current_sign_in_at" = ?,
"sign_in_count" = ?, "updated_at" = ? WHERE "users"."id" = 2  [["last_sign_in_at", Fri,
08 Nov 2013 14:13:56 UTC +00:00], ["current_sign_in_at", Fri, 08 Nov 2013 14:18:49 UTC
+00:00], ["sign_in_count", 3], ["updated_at", Fri, 08 Nov 2013 14:18:49 UTC +00:00]]
(143.6ms)  commit transaction
Redirected to http://localhost:3000/
Completed 302 Found in 239ms (ActiveRecord: 144.5ms | Search: 0.0ms)
```

**Comment by robinboening** [comment]
Had the same issue until adding an initializer that skips before action `verify_authenticity_token`.

@josevalim shouldn't this issue be opened again?

**Comment by latortuga** [comment]
Disabling `verify_authenticity_token` is a dangerous fix - it is putting out a fire by vacuuming up smoke. Please don't disable CSRF protection for your app.

If you have CSRF enabled and you see an authenticity token in the request, check that you actually have a session token in your cookies. Devise has exhaustive tests of this functionality - it is unlikely to be a devise issue.

**Comment by robinboening** [comment]
@latortuga: its not skipped for all requests in the app, but for our `UserSessionsController` that inherits from `Devise::SessionsController` only.

**Comment by latortuga** [comment]
If the session is empty, obviously the token will not be verified - this is how CSRF protection in Rails works! Check if you're using `httponly` or `secure` flags for your sessions and that they match your desired configuration. If you watch your cookies when you visit the site and you don't see your app's session key (`config/initializers/session_store.rb` `key` config), then you don't have a token that can validate against the CSRF token server-side - this will cause the issue.

If you're not validating CSRF protection for login/logout, a malicious attacker could cause you to logout or login as someone else without your knowledge.

**Comment by talal7860** [comment]
This solved my problem

I overrode the devise sessions controller by creating a new file in controller/users/ with the name of sessions_controller.rb
    class Users::SessionsController < Devise::SessionsController
        skip_before_filter :verify_authenticity_token, :only => [:destroy]
    end

**Comment by danielbonnell** [comment]
I've tried every solution on here to no avail. I am running Rails 4.2 and the latest version of Devise on Heroku. My issue only seems to occur in Firefox and the environment doesn't matter. It happens both in development, test, and production. I can get around it in testing by stubbing the user login, but I can't seem to get around it in the other environments. 

I can't figure out what caused the issue in the first place. I had no problems logging in through Firefox as recently as a week ago. I haven't pushed any new code up to the production app recently.

**Comment by talal7860** [comment]
Make sure that your csrf tag are being displayed as it should in firefox,
this solved my problem, I accidentally deleted the csrf tags

On Sunday, April 5, 2015, Daniel Bonnell <notifications@github.com
<javascript:_e(%7B%7D,'cvml','notifications@github.com');>> wrote:

> I've tried every solution on here to no avail. I am running Rails 4.2 and
> the latest version of Devise on Heroku. My issue only seems to occur in
> Firefox and the environment doesn't matter. It happens both in development,
> test, and production. I can get around it in testing by stubbing the user
> login, but I can't seem to get around it in the other environments.
>
> I can't figure out what caused the issue in the first place. I had no
> problems logging in through Firefox as recently as a week ago. I haven't
> pushed any new code up to the production app recently.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/plataformatec/devise/issues/2734#issuecomment-89655905>
> .
>


-- 
Thanks & Regards,
Talal Arshad

**Comment by danielbonnell** [comment]
I had thought of that, but the CSRF tags are present in the DOM. Turns out the issue (at least for me) was being caused by the [Blur / DoNotTrackPlus](https://addons.mozilla.org/en-US/firefox/addon/donottrackplus/) extension for Firefox. I had previously disabled it, along with my other extensions, but that had no effect. When I removed it, however, the problem went away. I further confirmed that by having a friend test the site on his machine using Firefox (without the extension).

Hope this helps someone in the future.

**Comment by pankajp20** [comment]
I'm getting same error while updating a record on IE only. This do not happen application wide but occur for specific action of a controller. While submitting a form (action is used to update a record in DB), received error "Can't verify CSRF token authenticity". When checked form in IE, creates hidden field for "authenticity_token" but submitting the form redirect me to root URL of app (in application controller) and gives same error in console. Also form in IE contains "csrf_meta_tags".

I tried to skip validating the CSRF token authenticity for this specific action but then form's data is not submitted (write skip_before_action :verify_authenticity_token, only: [:] in my controller).

This problem is coming after app migrated from Rails 3.2 to 4.2 using Devise 3.4. Any one have idea on this issue.

**Comment by latortuga** [comment]
I know this can be frustrating but it isn't helpful to bump this thread saying that you're experiencing this bug without a reproducible test case. If you have a sample application that exhibits the behavior, we can look at it and try to fix it. A good bug report will include 1. what went wrong 2. what you tried to fix it and 3. a way to reproduce the error.

**Comment by talal7860** [comment]
And please try devise by installing it on a fresh install of rails with the
same versions. You will figure out what are you doing wrong

On Monday, April 6, 2015, Drew Ulmer <notifications@github.com> wrote:

> I know this can be frustrating but it isn't helpful to bump this thread
> saying that you're experiencing this bug without a reproducible test case.
> If you have a sample application that exhibits the behavior, we can look at
> it and try to fix it. A good bug report will include 1. what went wrong 2.
> what you tried to fix it and 3. a way to reproduce the error.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/plataformatec/devise/issues/2734#issuecomment-90158302>
> .
>


-- 
Thanks & Regards,
Talal Arshad

**Comment by itsmechlark** [comment]
I found something, `csrf-token` in meta tag, form and session store is not the same.

meta tag: `wKPKeOu1qfe5XKyoNBaDrlWYEZH4MCISVK++nnoMq41i1QiGIr8vZ2bzcJL5fJC79LV9XkXrgklDdnVsBmMamA==`
form: `GvEP+PmPduvdI00DQ1m1N6XRF1slmpu6ibFAqyLx9uu4h80GMIXwewKMkTmOM6YiBPx7lJhBO+GeaItZXp5H/g==`
session_store: `onbC/skKhpDfr9w6zWoTFaEtbM+926BbF9nL8nxvsRU=`

with this configuration:
```ruby
Rails.application.config.session_store :active_record_store, key: 'SESSION', expire_after: 24.hours
ActionDispatch::Session::ActiveRecordStore.session_class = Session

class Session < ActiveRecord::SessionStore::Session
  belongs_to :user
  belongs_to :devise_session, dependent: :destroy
  belongs_to :devise_session_log

  before_save :associate_about_user

  def self.find_by_session_id(session_id)
    where(session_id: session_id).first
  end

  def warden_session
    data['warden.user.user.session']
  end

  def warden_key
    data['warden.user.user.key']
  end

  def associate_about_user
    self.user = User.find(warden_key[0][0]) if warden_key

    if warden_session
      unique_session_id = warden_session['unique_session_id']
      session = DeviseSession.find_by_unique_session_id(unique_session_id)
      self.devise_session = session if session

      token = warden_session['unique_auth_token_id']
      log = DeviseSessionLog.find_by_unique_auth_token_id(token)
      self.devise_session_log = log if log
    end
  end
end
```

rails 4.2.1
devise 3.4.1
activerecord-session_store 0.1.1
[devise_security_extension](https://github.com/itsmechlark/devise_security_extension)
evise-encryptable 0.2.0

**Comment by pankajp20** [comment]
Thank you @latortuga and @talal7860, but the actual problem was the "div"s. Structure of that page is based on "div" and one "div" (where the form lies) was placed inside table body mistakenly. And that "div" floats in the page causing this bug. Either no table should be used or the "div" should be inside a "td" tag.

**Comment by Aleksandaar** [comment]
If you are using forms and turbo-links, the issue can be solved by adding a `:authenticity_token => true`, like:
<%= form_for @your_model, :remote => true, :authenticity_token => true do |f| %>
<% end %>

**Comment by pdbradley** [comment]
Had this issue and fixed it by removing the :domain => :all 

```ruby
Myapp::Application.config.session_store :cookie_store, key: '_myapp_session',:domain => :all

to

Myapp::Application.config.session_store :cookie_store, key: '_myapp_session'
```

**Comment by 1dot44mb** [comment]
Make sure the browser you use accepts cookies (you can do it in settings). You can suspect this is the problem if the error happens in specific browsers (eg. Firefox).

**Comment by david-meza** [comment]
I had this issue only on heroku as well. Removing `domain: :all worked` for me

**Comment by antonkoh** [comment]
Having a similar problem on all browsers with my app:
https://github.com/antonkoh/railsrep/tree/master/lesson17/my_app

First way to reproduce:
1. As unauthorized user open any post on the main page.
2. In another tab sign in (test00@test.com / 12345678)
3. On the first tab, enter any symbols in the New Comment field and try to submit.

Second way to reproduce:
1. As an authorized user (test00@test.com / 12345678) open any post on the main page.
2. In another tab sign out.
3. On the first tab try to add a comment.

Both ways end in ActionController::InvalidAuthenticityToken exception.
This is also reproduced for any other POST requests from an unrefreshed page.

Devise 3.5.2, Rails 4.2.4

**Comment by drewhamlett** [comment]
Just in case anybody runs across this in Google(because I did).  I used the domain: :all flag, and sign_out stopped working on  production.  I had two cookie sessions and I guess devise was only clearing one?  Anyway removed domain: : all and everything is good again.

**Comment by eddroid** [comment]
I also resolved a similar Devise issue by unbundling the `rails-api` gem. Two different failures:

1. `authenticity_tokens` generated by `form_for` were invalid on `POST sessions#new`.
1. I had `before_filter :authenticate_user!` in `ApplicationController`. So there were no public pages (other than login). Even after a successful login `POST`, the session would remain empty and I'd be redirected to the login page again with a `401 Unauthorized`.

Theory: `rails-api` (or something) is disabling the session, even in non `ActionController:API`-subclassed controllers. Since the token in the session needs to match the token in the form, token verification fails. The [Rails source](https://github.com/rails/rails/blob/master/actionpack/lib/action_controller/metal/request_forgery_protection.rb#L368) seems to indicate that if a session token comes back empty, a new random one is generated. This makes sense when rendering a form on a `GET`, but is unnecessary and confusing when handling `POST`/`PUT`/etc. The error isn't that the token doesn't match. The error is that, while your form has a token, the session doesn't, and your form's token doesn't match the one Rails just randomly generated.

**Comment by allpumpedup** [comment]
@drewhamlett suggestion of removing domain: :all flag worked for me.

**Comment by adrianoresende** [comment]
I have exactly the same of the first post problem, after switching to rails 5.
It solves cases removing `rails-api` but It has included native for Rails 5 and how to solve? =(

Rails 5.0.0.beta3
Devise 4.0.0.rc2 (support for Rails 5)

My error:
```
Started POST "/admin/sign_in" for 127.0.0.1 at 2016-03-22 12:08:56 -0300
Processing by Devise::SessionsController#create as HTML
  Parameters: {"utf8"=>"✓", "authenticity_token"=>"c2r8dL0baHl3NLUpW9s3fMoTcxuDaezGwSD2+2B4l7MHJJ0Hnwl4KTuowz1ewwSk6Hklqb/N+W40YJAa6YEBzw==", "user"=>{"email"=>"zullcore@gmail.com", "password"=>"[FILTERED]", "remember_me"=>"0"}, "commit"=>"Entrar"}
  User Load (0.4ms)  SELECT  `users`.* FROM `users` WHERE `users`.`email` = 'zullcore@gmail.com' ORDER BY `users`.`id` ASC LIMIT 1
   (0.2ms)  BEGIN
  SQL (0.3ms)  UPDATE `users` SET `current_sign_in_at` = '2016-03-22 15:08:56', `sign_in_count` = 47, `updated_at` = '2016-03-22 15:08:56' WHERE `users`.`id` = 2
  SQL (0.4ms)  INSERT INTO `versions` (`item_type`, `item_id`, `event`, `object`, `created_at`, `object_changes`) VALUES ('User', 2, 'update', '---\nid: 2\nname: Adriano\nprofile: \nadmin: true\ncreated_at: 2016-03-21 23:51:39.000000000 Z\nupdated_at: 2016-03-22 15:08:49.000000000 Z\nemail: zullcore@gmail.com\nencrypted_password: \"$ ... \"\nreset_password_token: \nreset_password_sent_at: \nremember_created_at: \nsign_in_count: 46\ncurrent_sign_in_at: 2016-03-22 15:08:49.000000000 Z\nlast_sign_in_at: 2016-03-22 15:08:49.000000000 Z\ncurrent_sign_in_ip: 127.0.0.1\nlast_sign_in_ip: 127.0.0.1\nphoto_file_name: \nphoto_content_type: \nphoto_file_size: \nphoto_updated_at: \ncreated_by: \nupdated_by: \n', '2016-03-22 15:08:56', '---\ncurrent_sign_in_at:\n- 2016-03-22 15:08:49.000000000 Z\n- 2016-03-22 15:08:56.000000000 Z\nsign_in_count:\n- 46\n- 47\nupdated_at:\n- 2016-03-22 15:08:49.000000000 Z\n- 2016-03-22 15:08:56.000000000 Z\n')
   (47.5ms)  COMMIT

Can't verify CSRF token authenticity

   (0.1ms)  BEGIN
   (0.1ms)  COMMIT
   (0.1ms)  BEGIN
  SQL (0.3ms)  UPDATE `users` SET `last_sign_in_at` = '2016-03-22 15:08:56', `sign_in_count` = 48 WHERE `users`.`id` = 2
  SQL (0.3ms)  INSERT INTO `versions` (`item_type`, `item_id`, `event`, `whodunnit`, `object`, `created_at`, `object_changes`) VALUES ('User', 2, 'update', '2', '---\nencrypted_password: \"$ ... \"\nid: 2\ncurrent_sign_in_at: 2016-03-22 15:08:56.000000000 Z\nlast_sign_in_at: 2016-03-22 15:08:49.000000000 Z\ncurrent_sign_in_ip: 127.0.0.1\nlast_sign_in_ip: 127.0.0.1\nsign_in_count: 47\nname: Adriano\nprofile: \nadmin: true\ncreated_at: 2016-03-21 23:51:39.000000000 Z\nupdated_at: 2016-03-22 15:08:56.000000000 Z\nemail: zullcore@gmail.com\nreset_password_token: \nreset_password_sent_at: \nremember_created_at: \nphoto_file_name: \nphoto_content_type: \nphoto_file_size: \nphoto_updated_at: \ncreated_by: \nupdated_by: \n', '2016-03-22 15:08:56', '---\nlast_sign_in_at:\n- 2016-03-22 15:08:49.000000000 Z\n- 2016-03-22 15:08:56.000000000 Z\nsign_in_count:\n- 47\n- 48\n')
   (0.4ms)  COMMIT
Redirected to http://localhost:3000/admin
Completed 302 Found in 186ms (ActiveRecord: 50.2ms)

Started GET "/admin" for 127.0.0.1 at 2016-03-22 12:08:56 -0300
Processing by Admin::PagesController#dashboard as HTML
Redirected to http://localhost:3000/admin/login
Filter chain halted as :require_login rendered or redirected
Completed 302 Found in 2ms (ActiveRecord: 0.0ms)
```

**Comment by chris357** [comment]
Rails 4 on chrome.

**Findings:**
1. Issue only occurs if user logs in, logs out, and then tries to log in again.
2. Clearing the cache (shift + cmd + r) fixes the issues temporarily.

Maybe is something not being fully cleared on logout?

**Comment by ACPK** [comment]
@adrianoresende I'm also hitting this issue with Rails5. Did you find a solution?

**Comment by adrianoresende** [comment]
@ACPK I not found solution =/ 
I went back to the latest version Rails4

**Comment by adrianoresende** [comment]
@josevalim Why you do not reopen this issue? Because Rails5

**Comment by codeodor** [comment]
I'm also seeing it in Rails 5, for what it's worth.

**Comment by codeodor** [comment]
For those having issues on Rails 5, see https://github.com/plataformatec/devise/pull/4033/files

**Comment by vickodin** [comment]
In my case, all working fine (nginx + puma + rails 5 with turbolinks 5), even without previous advice.
But after ssl certificate installation, I have got the same error: Can't verify CSRF token authenticity.

I added the next line to the nginx config, as solution:
proxy_set_header X-Forwarded-Ssl     on;

And have no errors now.
I used devise, but the problem is not in devise. The usual forms was also broken.

**Comment by lazybios** [comment]
@vickodin  I have same problem even if I remove devise from gemfile.  Thank you for you tips about nginx.

**Comment by swaathi** [comment]
Facing the same issue with Rails 5 and multiple subdomains. I've also added the ```protect_from_forgery with: :exception, prepend: true``` line, but no go.

Since I'm using multiple subdomains, I've also added this to `session_store.rb`,

```Rails.application.config.session_store :cookie_store, {
  :key => '_ABC_session',
  :domain => {
  	production: '.abc.in',
  	development: '.lvh.me'
	}.fetch(Rails.env.to_sym, :all)
}```

Same codebase works with Rails 4. 😫 

Can anyone point me in the right direction?

**Comment by elkhan** [comment]
My issue with CSRF token related to paper_trail gem. There is a before_action :set_paper_trail_whodunnit I moved it below protect_from_forgery and configure_permitted_parameters then it started working. Hope it helps someone

**Comment by rdetert** [comment]
I was seeing this issue just now and it was related to me using Heroku + Cloudflare. I had the option set for "Enforce HTTPS for this URL" on Cloudflare. I've turned it off and it now works, but now I have to figure out what I'm going to do about SSL.

**Comment by holoiii** [comment]
@rdetert Were you able to figure out a solution? I am running into something similar - Heroku + Cloudflare (with their free flexible SSL) causes this issue.

**Comment by rdetert** [comment]
@holoiii Yes, though I didn't spend a ton of time testing different scenarios. What I ended up doing was this:
1) Add the Expedited SSL Add On from Heroku
2) Change Cloudflare SSL from Flexible to Full
3) Force SSL in ApplicationController (  force_ssl if: :ssl_configured? )

Heroku cannot use apex domains so Expedited SSL will make you to use https://www.mydomain.com, however, running with Cloudflare, it works fine using https://mydomain.com

Seems to work.

**Comment by ayghor** [comment]
+1 for

`proxy_set_header X-Forwarded-Ssl on;`

on rails 5 with nginx and ssl. Note that it worked for some people, whiel other would get `authenticity_token` error. And i was getting strange redirection behaviour too1

I don't use devise btw...

**Comment by mmatusSVK** [comment]
Thanks to vickodin,
for **rails 5** with **nginx** is needed not only to prepend` protect_from_forgery`, but also set up 
`proxy_set_header X-Forwarded-Ssl on;` in nginx site config.

**Comment by unkrich** [comment]
I was able to solve this problem after quite a lot of troubleshooting. I am using cloudflare with heroku. The fix was to switch Cloudflare's SSL from Flexible to Full. Then everything worked. That simple.
Note that I am using the 'beta'-ish feature of Heroku's paid dynos which allow you to upload an SSL certificate for free rather than the SSL Endpoint plugin (if this matters).

**Comment by corps** [comment]
Another potential cause worth mentioning: sessions not saving.  I had a problem in which my form was generating a csrf token, and it was being sent and received by the server correctly, but it was never matching the token in the session because a new one was being regenerated every request due to a session issue.

**Comment by santhoshcd** [comment]
I have the same problem but I am not using the device.
In application controller added 'protect_from_forgery' to fix CSRF issue.

To reproduce  
Step 1 : Login Page opened in 2 tabs (Tab1 and Tab2) 
Step 2:  From tab1 logged in  and logged out
Step 3: From tab2 I tried to log in then I got below because token destroyed from tab1 logout request
```

Processing LoginController#login (for 70.71.248.213 at 2016-11-05 14:37:03) [POST]
  Parameters: {"authenticity_token"=>"sobCryvXTgoi29Ifwb88nJicHY+3jmiybWvXWwfghQw=", "commit"=>"Login", "controller"=>"login", "password"=>"[FILTERED]", "os_info"=>"Windows", "action"=>"login", "username"=>"test", "browser_info"=>"Mozilla an unknown version"}

ActionController::InvalidAuthenticityToken
```
The Same issue is happening for Ajax post request in another tab3 (opened  another page before logout from tab1).

Do we have any solution to fix this issue?

**Comment by muhammadn** [comment]
I have this problem but `proxy_set_header X-Forwarded-Ssl on;` is a partial solution at least when i am using facebook OAuth login with Omniauth (and devise)

The solution is to use `proxy_set_header X-Forwarded-Proto $scheme;` instead of `proxy_set_header X-Forwarded-Ssl on;`

**Comment by rusllonrails** [comment]
@pdbradley thanks bro, your suggestion works perfect!

**Comment by paulewetzel** [comment]
Yep,   **proxy_set_header X-Forwarded-Proto $scheme;** worked for me as well. I was just about to throw a TV out a 2nd story window, and now I can watch it instead....thanks! 
location @unicorn {
    proxy_pass http://unicorn;
    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
    proxy_set_header Host $http_host;
    **proxy_set_header X-Forwarded-Proto $scheme;**
    proxy_redirect off;
  }

**Comment by janospapp** [comment]
Thank you guys! I've been struggling for days with authenticity token failure in chrome (in firefox it worked properly). After adding "proxy_set_header X-Forwarded-Proto $scheme;" to nginx config the problem is solved. I might messed up the config file when setting up ssl. Big thanks!!

**Comment by jarvisjohnson** [comment]
Following on from Cloudflare SSL comments above (@rdetert).  If you are not using Heroku, so you aren't able to access an automatic SSL addon, you need to add an actual certificate on your nginx server manually.

Here's how I set it up on Digital Ocean:

Generate a TLS cert from cloudflare (free, just below the SSL flexible/full/strict area).

On the server (as root):
 - add the certs: {app_name}_cert.cert & {app_name}_key.key to /home/{your-app-name}
 - Change /etc/nginx/sites-available/{your-app-name} to include:

```
server {
        listen 80;
        listen 443;

        ssl     on;
        ssl_certificate         /home/{app_name}/{app_name}_cert.pem;
        ssl_certificate_key     /home/{app_name}/{app_name}_key.key;
        ...
}
```

Set Cloudflare SSL settings to Full(strict) to force Cloudflare to use your server certs

Finally, configure `config/environments/production.rb` : `config.force_ssl = true`

**Comment by Raja-Sharma** [comment]
Rails 5.1, aws elastic beanstalk with cloudfront.

Had the same error and tried everything here to no avail. Fixed it by setting "Forward Cookies" to "all" from the default "none (Improves Caching)" in Distribution Settings > Behavior > (select behavior) > Edit.

**Comment by inem** [comment]
I recently caught the same problem.

This commit caused it: 
![image](https://user-images.githubusercontent.com/9665/35100165-05ec0fd6-fc5b-11e7-9f13-44976dcdb04d.png)

Looks fine, but turns out it is a bad idea to access `current_user` in `before_action` in ApplicationController

Devise controllers are still controllers, and are inherited from ApplicationController.

So I tried to access `current user` in before_action, before devise finished its `protect_from_forgery` stuff.

The fix was to put `protect_from_forgery` before my action

**Comment by lostphilosopher** [comment]
I had this problem, for me it was nothing to do with Devise. It was Rails5 + Heroku + CloudFlare as explained here: http://til.obiefernandez.com/posts/875a2a69af-cloudflare-flexible-ssl-mode-breaks-rails-5-csrf. 

> "Turn Cloudflare SSL to Full mode. Problem solved."

**Comment by czeisser** [comment]
I was experiencing this problem too. It turned out that the `prepend: true` option for `protect_from_forgery`  was being overwritten. 
I was using a gem that also declared `protect_from_forgery` but without the `prepend` option so it defaulted to false. (In my case auto_session_timeout gem was the culprit.)
To fix this I moved my `protect_from_forgery` after the `auto_session_timeout` in the ApplicationController. 

```ruby 
  class ApplicationController < ActionController::Base
  ...
  
  auto_session_timeout Rails.configuration.session_timeout_length
  
  #has to be after auto_session_timeout so that prepend will not be overwritten.
  protect_from_forgery with: :exception, prepend: true
```

**Comment by ruvaleev** [comment]
For me the problem was solved when wrote this in my application_controller.rb

class ApplicationController < ActionController::Base
  protect_from_forgery with: :exception
  check_authorization :unless => :devise_controller?

Another order of strings above was the reason of exception. Hope it'll help somebody.

**Comment by GuanglinDu** [comment]
I ran across this problem after I switched back and forth between the development and production environment on the same machine with browser Firefox. The 2 data sets are quite different. It took me 2 days to find out that it was caused by the Firefox cache. After the cache was cleared, Rails (4.2.11.1 & Ruby 2.2.5) functions properly. The lesson learned: never mix up  the development and production environment on the same machine.

**Comment by clst** [comment]
I also came across this problem. After verifying that it worked fine on all other machines I tried it I got suspicious.

Turns out Server 2016's Internet Explorer needs to have the Rails site added to "Trusted Sites" then it magically starts working. I have not investigated further.

**Comment by cruznunez** [comment]
i came across the problem and had nothing to do with Devise. My solution was different than the ones here. 

I have two modal forms. It seemed like the 1st one was working and the 2nd one wasn't. I checked the authenticity tokens on them and they are different per form. I grabbed the authenticity token from the metatag and made that the value of the authenticity token hidden field for the 2nd form.

Tried it and it fixed it. Here's the coffescript:

```coffee
editNoteModal = ->
  $('[href="#edit-note"]').click ->
    # code

    modal = $ '#edit-note'

    # code
    # code
    # code

    token = $('meta[name="csrf-token"]').attr('content')
    authTokenField = modal.find '[name=authenticity_token]'
    authTokenField.val token

$(window).on 'turbolinks:load', editNoteModal
```

**Comment by camilova** [comment]
This error was the **browser's cache** for me. I was working on two different projects, and both were on `localhost:3000`. When I came back to the first project after work on the another, I could not log in any more due **wrong verification token**, then I looked on the Chrome resources and there was a coockie from the another project. I have removed the wrong cookie, but it did not work, then I delete all the data of the browser through its configuration menu, and all is working get!

Before my solution I tried with a private window, on wich all worked, so I came to the idea of the browser's cache, that finally was the solution.

**Comment by Arish-theGalaxyBea** [comment]
The simplest solution for the problem is do standard things in your controller or you can directly put it into **ApplicationController**

`class ApplicationController < ActionController::Base`
    ` protect_from_forgery with: :exception, prepend: true`
`end`

**Comment by schappim** [comment]
I had gone from "CSRF was previously working on localhost and production" to suddenly not working w/ ngrok.

If you've using subdomains (including ngrok which uses `[yoursubdomain].[country].ngrok.io` , make sure that you have updated your session store to have a `tld_length` that matches the length of your subdomains.

Eg:

```ruby
Rails.application.config.session_store :cookie_store, key: "_example_session", domain: :all, tld_length: 3
```

For more info see: [https://github.com/heartcombo/devise/wiki/How-To:-Use-subdomains](https://github.com/heartcombo/devise/wiki/How-To:-Use-subdomains#)

**Comment by lapser** [comment]
In my case I get the "Can't verify CSRF token authenticity" log message after Devise form submission when I use [this code](https://guides.rubyonrails.org/action_controller_overview.html#default-url-options) in `app/controllers/application_controller.rb`.

```
class ApplicationController < ActionController::Base
  def default_url_options
    { locale: I18n.locale }
  end
end
```

I have to submit Devise forms two times to get e.g. signed in or receive errors. It also seems that this happens randomly.

If I remove the code, Devise seems to work as expected.

Using Rails 7.0.4.2 and Devise 4.9.0

**Comment by codeodor** [comment]
I ran into this and it was due to me having embedded one `form` tag inside another so that the app was expecting one token but getting the other. The solution was to ensure I closed out the first form tag before opening the second.

**Comment by lapser** [comment]
> P.S.: Maybe this is being solved by #5567

I tryed the recent rails 7.0.4.3 branch from @carlosantoniodasilva (https://github.com/heartcombo/devise/pull/5567) and the issue related to the "Can't verify CSRF token authenticity" [while using `default_url_options`](https://github.com/heartcombo/devise/issues/2734#issuecomment-1466850562) seems not be solved.

**Comment by lapser** [comment]
> If you are using forms and turbo-links, the issue can be solved by adding a `:authenticity_token => true`, like:
> 
> ```
> <%= form_for @your_model, :remote => true, :authenticity_token => true do |f| %>
> <% end %>
> ```

The above works to avoid getting the "Can't verify CSRF token authenticity" message using rails 7.0.4.3 and devise 4.9.0 with Hotwire Turbo (without `:remote => true`). It's the same as adding `<%= hidden_field_tag :authenticity_token, form_authenticity_token %>` into your form, like this:

```
<%= form_for @your_model, url: ... do |f| %>
   <%= hidden_field_tag :authenticity_token, form_authenticity_token %>
<% end %>
```
